fl4health/clients/basic_client.py

```diff
@@ -44,30 +48,150 @@ def __init__(
             checkpointer (Optional[TorchCheckpointer], optional): Checkpointer to be used for client-side
                 checkpointing. Defaults to None.
         """
-        super().__init__(data_path, device)
+        self.data_path = data_path
+        self.device = device
         self.metrics = metrics
         self.checkpointer = checkpointer
+
+        self.client_name = self.generate_hash()
+        self.initialized = False  # Whether or not the client has been setup
+        self.model_weights_initialized = False
+
+        # Loss and Metric management
         self.train_loss_meter = LossMeter.get_meter_by_type(loss_meter_type)
         self.val_loss_meter = LossMeter.get_meter_by_type(loss_meter_type)
-
         self.train_metric_manager = MetricManager(metrics=self.metrics, metric_manager_name="train")
         self.val_metric_manager = MetricManager(metrics=self.metrics, metric_manager_name="val")
 
+        # Optional variable to store the weights that the client was initialized with during each round of training
+        self.initial_weights: Optional[NDArrays] = None
+
+        self.wandb_reporter: Optional[ClientWandBReporter] = None
+        self.total_steps: int = 0  # Need to track total_steps across rounds for WANDB reporting
+
+        # Attributes to be initialized in setup_client
+        self.parameter_exchanger: ParameterExchanger
         self.model: nn.Module
         self.optimizer: torch.optim.Optimizer
-
         self.train_loader: DataLoader
         self.val_loader: DataLoader
         self.num_train_samples: int
         self.num_val_samples: int
         self.learning_rate: float
 
-        # Need to track total_steps across rounds for WANDB reporting
-        self.total_steps: int = 0
+    def _maybe_checkpoint(self, comparison_metric: float) -> None:
+        """
+        If checkpointer exists, maybe checkpoint model based on the comparison metric.
+
+        Args:
+            comparison_metric (float): The metric value obtained by the current model.
+                Used to decide if to checkpoint model by checkpointer.
+        """
+        if self.checkpointer:
+            self.checkpointer.maybe_checkpoint(self.model, comparison_metric)
+
+    def generate_hash(self, length: int = 8) -> str:
+        """
+        Generates unique hash used as id for client.
+
+        Args:
+           length (int): The length of the hash.
+
+        Returns:
+            str: client id
+        """
+        return "".join(random.choice(string.ascii_lowercase) for i in range(length))
+
+    def get_parameters(self, config: Config) -> NDArrays:
+        """
+        Determines which weights are sent back to the server for aggregation. This uses a parameter exchanger to
```

Comment:
Super minor: maybe we can say "parameters" instead of "weights" in the "determines which weights..." since weights sort of implies we only exchange model weights?

Reply:
Good call!